scipio/src/shared.rs

```diff
@@ -0,0 +1,344 @@
+    /// [`Shared`]: struct.Shared.html
+    /// [`None`]: https://doc.rust-lang.org/beta/std/option/enum.Option.html#variant.None
+    /// [`WeakShared::new`]: struct.WeakShared.html#method.new
+    pub fn upgrade(&self) -> Option<Shared<T>> {
+        self.inner.upgrade().map(|inner| Shared { inner })
+    }
+
+    /// Returns `true` if the two WeakShared point to the same allocation (similar to [`ptr::eq`])
+    ///
+    /// [`ptr::eq`]: https://doc.rust-lang.org/beta/std/ptr/fn.eq.html
+    pub fn ptr_eq(&self, other: &WeakShared<T>) -> bool {
+        self.inner.ptr_eq(&other.inner)
+    }
+
+    /// Get the number of strong pointers to this allocation
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use scipio::{Shared, WeakShared};
+    ///
+    /// let x = Shared::new(0);
+    /// assert_eq!(Shared::strong_count(&x), 1);
+    /// let y = Shared::downgrade(&x);
+    /// assert_eq!(y.strong_count(), 1);
+    /// ```
+    pub fn strong_count(&self) -> usize {
+        self.inner.strong_count()
+    }
+
+    /// Get the number of weak pointers to this allocation
+    ///
+    /// If no strong pointers remain, this will return zero.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use scipio::{Shared, WeakShared};
+    ///
+    /// let x = Shared::new(0);
+    /// assert_eq!(Shared::weak_count(&x), 0);
+    /// let y = Shared::downgrade(&x);
+    /// assert_eq!(y.weak_count(), 1);
+    /// ```
+    pub fn weak_count(&self) -> usize {
+        self.inner.weak_count()
+    }
+}
+
+impl<T> Shared<T> {
+    /// Creates a new `Shared` object
+    ///
+    /// This shared object implements [`Clone`] and is reference counted so it is safe to pass
+    /// to multiple tasks in the same [`LocalExecutor`]
+    ///
+    ///
+    /// # Examples
+    /// ```
+    /// use scipio::Shared;
+    ///
+    /// let x = Shared::new(0);
+    /// ```
+    /// [`Clone`]: https://doc.rust-lang.org/std/clone/trait.Clone.html
+    /// [`LocalExecutor`]: struct.LocalExecutor.html
+    pub fn new(t: T) -> Self {
+        Shared {
+            inner: Rc::new(UnsafeCell::new(t)),
+        }
+    }
+
+    /// Returns `true` if the two `Shared` point to the same allocation (similar to [`ptr::eq`])
+    ///
+    /// [`ptr::eq`]: https://doc.rust-lang.org/beta/std/ptr/fn.eq.html
+    pub fn ptr_eq(this: &Shared<T>, other: &Shared<T>) -> bool {
+        Rc::ptr_eq(&this.inner, &other.inner)
+    }
+
+    /// Executes the closure `func`, with a mutable reference to this `Shared`'s inner value as a
+    /// parameter
+    ///
+    /// # Examples
+    /// ```
+    /// use scipio::Shared;
+    ///
+    /// let x = Shared::new(0);
+    /// x.do_with(|x| assert_eq!(*x, 0) );
+    /// x.do_with(|x| *x = 2 );
+    /// x.do_with(|x| assert_eq!(*x, 2) );
+    /// ```
+    pub fn do_with<F, R>(&self, func: F) -> R
+    where
+        F: FnOnce(&mut T) -> R,
+        R: Sized,
+    {
+        let k = unsafe { &mut *self.inner.get() };
+        func(k)
+    }
```

Comment:
This API is fundamentally unsound due to re-entrancy:

```rust
#[test]
fn unsound() {
    let cell = Shared::new("hello world".to_string());
    cell.do_with(|r1| {
        let s = r1.as_str();
        cell.do_with(|r2| {
            *r2 = String::new();
            eprintln!("{:?}", s)
        })
    })
}
```

I *think* this fundamentally requires `RefCell`-like runtime flag to make it sound.

Rule-of-thumb: be extra careful if calling user-supplied code (`FnOnce`) near `unsafe`.

Reply:
Minor addendum: idiomatic name would be just `with`: https://doc.rust-lang.org/stable/std/thread/struct.LocalKey.html#method.with

Reply:
If this requires a `RefCell`-like runtime flag to work, then it is a pointless API since the whole idea of this is to experiment with ways to do away with the `RefCell` runtime check =)

If there is no way to prevent reentrancy, I agree that your example provides an example of something that could break this.

This leaves us with the other methods in `Shared`: instead of exposing `with` we can use that internally to update a value and return a value if the type implements `Copy`. (`swap` and `inner`)

I don't see a lot of upside in `inner` returning the value since it is already possible to do that without a `RefCell`. But 
`swap` still seems to me like a useful API as it allows you to mutate the contents in a controlled way (and it actually doesn't even need `Copy` as it is just moving the ownership)

So I think the direction I am going with this will probably be making `with` private and exposing just `swap` and `inner`.